src/routing_table.rs

```diff
@@ -0,0 +1,205 @@
+                let section: Vec<_> = section
+                    .member_nodes()
+                    .filter(|node| node.name() != our_id.name())
+                    .cloned()
+                    .collect();
+                let dg_size = section.len();
+                return Ok((section, dg_size));
+            }
+
+            candidates(target_name, our_id, sections)?
+        }
+        DstLocation::Prefix(prefix) => {
+            if prefix.is_compatible(sections.our().prefix())
+                || prefix.is_neighbour(sections.our().prefix())
+            {
+                // only route the message when we have all the targets in our chain -
+                // this is to prevent spamming the network by sending messages with
+                // intentionally short prefixes
+                if prefix.is_compatible(sections.our().prefix())
+                    && !prefix.is_covered_by(sections.prefixes())
+                {
+                    return Err(RoutingError::CannotRoute);
+                }
+
+                let is_compatible = |(pfx, section)| {
+                    if prefix.is_compatible(pfx) {
+                        Some(section)
+                    } else {
+                        None
+                    }
+                };
+
+                // Exclude our name since we don't need to send to ourself
+                let targets: Vec<_> = sections
+                    .all()
+                    .filter_map(is_compatible)
+                    .flat_map(EldersInfo::member_nodes)
+                    .filter(|node| node.name() != our_id.name())
+                    .cloned()
+                    .collect();
+                let dg_size = targets.len();
+                return Ok((targets, dg_size));
+            }
+
+            candidates(&prefix.lower_bound(), our_id, sections)?
+        }
+        DstLocation::Direct => return Err(RoutingError::CannotRoute),
+    };
+
+    Ok((best_section, dg_size))
+}
+
+// Obtain the delivery group candidates for this target
+fn candidates(
+    target_name: &XorName,
+    our_id: &PublicId,
+    sections: &SectionMap,
+) -> Result<(Vec<P2pNode>, usize)> {
+    let filtered_sections = sections
+        .sorted_by_distance_to(target_name)
+        .into_iter()
+        .map(|(prefix, members)| (prefix, members.len(), members.member_nodes()));
+
+    let mut dg_size = 0;
+    let mut nodes_to_send = Vec::new();
+    for (idx, (prefix, len, connected)) in filtered_sections.enumerate() {
+        nodes_to_send.extend(connected.cloned());
+        dg_size = delivery_group_size(len);
+
+        if prefix == sections.our().prefix() {
+            // Send to all connected targets so they can forward the message
+            nodes_to_send.retain(|node| node.name() != our_id.name());
+            dg_size = nodes_to_send.len();
+            break;
+        }
+        if idx == 0 && nodes_to_send.len() >= dg_size {
+            // can deliver to enough of the closest section
```

Comment:
not quite understand this comment.
does it want to say:
Closest section already having enough candidates to deliver ?

Reply:
To be honest I'm not sure what it the comment trying to say myself. I only copied this code from elsewhere. My suspicion is that this whole function might not be needed because we don't have the concept of "being connected" anymore. I'd come back to this in some other PR.

Reply:
Ok. just to mark this

Reply:
I think what we should do is to write some (at least basic) unit tests for this module and then refactor and simplify it. Something I'd do in a different PR though.